src/colmap/util/threading_ctrl.h

```diff
@@ -0,0 +1,123 @@
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include "colmap/util/logging.h"
+#include "colmap/util/threading.h"
+#include "colmap/util/timer.h"
+
+#include <atomic>
+#include <climits>
+#include <functional>
+#include <future>
+#include <list>
+#include <queue>
+#include <thread>
+#include <type_traits>
+#include <unordered_map>
+
+namespace colmap {
+
+// Reimplementation of threading with thread-related functions outside
+// controller Following util/threading.h
+// BaseController that supports templating in ControllerThread
+class BaseController {
+ public:
+  BaseController();
+  virtual ~BaseController() = default;
+
+  // Set callbacks that can be triggered within the main run function.
+  void AddCallback(int id, const std::function<void()>& func);
+
+  // This is the main run function to be implemented by the child class. If you
+  // are looping over data and want to support the pause operation, call
+  // `BlockIfPaused` at appropriate places in the loop. To support the stop
+  // operation, check the `IsStopped` state and early return from this method.
+  virtual void Run() = 0;
+
+  // check if the thread is stopped
+  void SetCheckIfStoppedFunc(const std::function<bool()>& func);
+  bool CheckIfStopped();
+
+ protected:
+  // Register a new callback. Note that only registered callbacks can be
+  // set/reset and called from within the thread. Hence, this method should be
+  // called from the derived thread constructor.
+  void RegisterCallback(int id);
+
+  // Call back to the function with the specified name, if it exists.
+  void Callback(int id) const;
+
+ private:
+  // list of callbacks
+  std::unordered_map<int, std::list<std::function<void()>>> callbacks_;
+  // check_if_stop function
+  std::function<bool()> check_if_stopped_fn_;
+};
+
+// Helper class to create single threads with simple controls
+// Similar usage as ``Thread`` class in util/threading.h except for
+// initialization. e.g.,
+//
+// std::shared_ptr<Controller> controller = std::make_shared<Controller>(args);
+// std::unique_ptr<ControllerThread<Controller>> thread =
+// std::make_unique<ControllerThread<Controller>>(controller);
+//
+//
+template <class Controller>
+class ControllerThread : public Thread {
+  // check if the Controller class is inherited from BaseController
+  static_assert(std::is_base_of<BaseController, Controller>::value,
+                "The controller needs to be inherited from BaseController");
+
+ public:
+  explicit ControllerThread(std::shared_ptr<Controller> controller) {
+    controller_ = controller;
+    controller_->SetCheckIfStoppedFunc([&]() { return IsStopped(); });
+  }
+  ~ControllerThread() = default;
+
+  // do BlockIfPaused() every time before checking IsStopped()
+  bool IsStopped() {
+    BlockIfPaused();
+    return Thread::IsStopped();
+  }
+
+  // add a controller-specific callback
+  void AddControllerCallback(int id, const std::function<void()>& func) {
+    controller_->AddCallback(id, func);
+  }
```

Comment:
I would prefer exposing `controller_` as const with a getter such that we're not tied to the interface of `AddCallback` (which might change in the future).

Reply:
Nice idea. This is done.